File: code/dynamic_programming/src/edit_distance/edit_distance.java
// dynamic programming | edit distance | Java
// Part of Cosmos by OpenGenus Foundation

import java.util.Scanner;

public class edit_distance {
    public static int edit_DP(String s,String t){
        int l1 = s.length();
        int l2 = t.length();
        
        int dp[][] = new int[l1+1][l2+1];

        for(int i=0; i<=l2; i++)
            dp[0][i] = i;

        for(int i=0; i<=l1; i++)
            dp[i][0] = i;

        for(int i=1; i<=l1; i++){
            for(int j=1; j<=l2; j++){
                if(s.charAt(l1-i) == t.charAt(l2-j))
                    dp[i][j] = dp[i-1][j-1];
                else
                    dp[i][j] = 1 + Math.min( dp[i-1][j-1] , Math.min (dp[i-1][j] , dp[i][j-1]) );
            }
        }
        return dp[l1][l2];
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        
        System.out.print("Enter first string >> ");
        String a = sc.next();

        System.out.print("Enter second string >> ");
        String b = sc.next();
        
        System.out.println("Edit Distance : " + edit_DP(a,b));
    }
}

File: code/dynamic_programming/src/edit_distance/edit_distance.go
/*
 * dynamic programming | edit distance | Go
 * Part of Cosmos by OpenGenus Foundation
 *
 * Compile: go build edit_distance.go
 */

package main

import (
	"fmt"
	"os"
)

func main() {

	if len(os.Args) != 3 {
		fmt.Println("Wrong input, write input next to the program call")
		fmt.Println("Example: ./edit_distance string1 string2")
		return
	}
	str1 := os.Args[1]
	str2 := os.Args[2]
	//str1 := "sunday"
	//str2 := "saturday"
	min := editDistance(str1, str2, len(str1), len(str2))
	fmt.Println(min)

}

func min(a, b int) int {
	if a < b {
		return a
	}

	return b
}

func editDistance(str1 string, str2 string, m int, n int) int {

	// Create table to store results of DP subproblems
	store := make([][]int, m+1)
	for i := range store {
		store[i] = make([]int, n+1)
	}
	for i := 0; i <= m; i++ {
		for j := 0; j <= n; j++ {
			if i == 0 {
				// When i is zero we must insert all (j) characters to get the second string
				store[i][j] = j
			} else if j == 0 {
				// Remove all characters in first string
				store[i][j] = i
			} else if str1[i-1] == str2[j-1] {
				// Last characters are equal, make recursion without last char
				store[i][j] = store[i-1][j-1]
			} else {
				// find minimum over all possibilities
				store[i][j] = 1 + min(min(store[i][j-1], store[i-1][j]), store[i-1][j-1])
			}
		}
	}
	return store[m][n]
}
